network/peer-monitoring-service/client/src/peer_states/network_info.rs

```diff
@@ -0,0 +1,171 @@
+#[derive(Clone, Debug)]
+pub struct NetworkInfoState {
+    base_config: BaseConfig, // The base config of this node
+    network_monitoring_config: NetworkMonitoringConfig, // The config for network monitoring
+    recorded_network_info_response: Option<NetworkInformationResponse>, // The last network info response
+    request_tracker: Arc<RwLock<RequestTracker>>, // The request tracker for network info requests
+}
+
+impl NetworkInfoState {
+    pub fn new(node_config: NodeConfig, time_service: TimeService) -> Self {
+        let base_config = node_config.base;
+        let network_monitoring_config = node_config.peer_monitoring_service.network_monitoring;
+        let request_tracker = RequestTracker::new(
+            network_monitoring_config.network_info_request_interval_ms,
+            time_service,
+        );
+
+        Self {
+            base_config,
+            network_monitoring_config,
+            recorded_network_info_response: None,
+            request_tracker: Arc::new(RwLock::new(request_tracker)),
+        }
+    }
+
+    /// Records the new network info response for the peer
+    pub fn record_network_info_response(
+        &mut self,
+        mut network_info_response: NetworkInformationResponse,
+        peer_network_id: &PeerNetworkId,
+        peer_metadata: PeerMetadata,
+    ) {
+        // Update the request tracker with a successful response
+        self.request_tracker.write().record_response_success();
+
+        // Sanity check the response depth from the peer metadata
+        let network_id = peer_network_id.network_id();
+        let is_valid_depth = match network_info_response.distance_from_validators {
+            0 => {
+                // Verify the peer is a validator and has the correct network id
+                let peer_is_validator = peer_metadata.get_connection_metadata().role.is_validator();
+                let peer_has_correct_network = match self.base_config.role {
+                    RoleType::Validator => network_id.is_validator_network(), // We're a validator
+                    RoleType::FullNode => network_id.is_vfn_network(),        // We're a VFN
+                };
+                peer_is_validator && peer_has_correct_network
+            },
+            1 => {
+                // Verify the peer is a VFN and has the correct network id
+                let peer_is_vfn = peer_metadata.get_connection_metadata().role.is_vfn();
+                let peer_has_correct_network = match self.base_config.role {
+                    RoleType::Validator => network_id.is_vfn_network(), // We're a validator
+                    RoleType::FullNode => network_id.is_public_network(), // We're a PFN
+                };
+                peer_is_vfn && peer_has_correct_network
+            },
+            distance_from_validators => {
+                // The depth must be less than or equal to the max
+                distance_from_validators <= MAX_DISTANCE_FROM_VALIDATORS
```

Comment:
Where is this being enforced?

Reply:
The server-side returns the `MAX_DISTANCE_FROM_VALIDATORS` if the node isn't connected to the network. But, to handle malicious peers, we also sanity check here so that invalid responses are limited to `MAX_DISTANCE_FROM_VALIDATORS`. Does that make sense?